sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -405,6 +409,246 @@ protected EventProcessor()
+
+                        // Allow the core dispatching loop to apply an additional set of retries over any provided by the consumer
+                        // itself, as a processor should be as resilient as possible and retain partition ownership if processing is
+                        // able to make forward progress.  If the retries are exhausted or a non-retriable exception occurs, the
+                        // consumer will be considered invalid and potentially refreshed.
+
+                        while (!cancellationSource.IsCancellationRequested)
+                        {
+                            try
+                            {
+                                eventBatch = await consumer.ReceiveAsync(EventBatchMaximumCount, Options.MaximumWaitTime, cancellationSource.Token).ConfigureAwait(false);
+                                await ProcessEventBatchAsync(partition, eventBatch, Options.MaximumWaitTime.HasValue, cancellationSource.Token).ConfigureAwait(false);
+
+                                // If event batches are successfully processed, then the need for forward progress is
+                                // satisfied, and the failure counts should reset.
+
+                                failedAttemptCount = 0;
+                                failedConsumerCount = 0;
+                            }
+                            catch (TaskCanceledException) when (cancellationSource.IsCancellationRequested)
+                            {
+                                // Do not log; this is an expected scenario when partition processing is asked to stop.
+
+                                throw;
+                            }
+                            catch (Exception ex) when (!(ex is DeveloperCodeException))
+                            {
+                                _ = OnProcessingErrorAsync(ex, partition, Resources.OperationReadEvents, CancellationToken.None);
+
+                                Logger.EventProcessorPartitionProcessingError(partition.PartitionId, Identifier, EventHubName, ConsumerGroup, ex.Message);
+                                retryDelay = retryPolicy.CalculateRetryDelay(ex, ++failedAttemptCount);
+
+                                if (!retryDelay.HasValue)
+                                {
+                                    // If the exception should not be retried, then allow it to pass to the outer loop; this is intended
+                                    // to prevent being stuck in a corrupt state where the consumer is unable to read events.
+
+                                    throw;
+                                }
+
+                                await Task.Delay(retryDelay.Value, cancellationSource.Token).ConfigureAwait(false);
+                            }
+                        }
+                    }
+                    catch (OperationCanceledException ex)
```

Comment:
There's a specific connotation to the `OperationCanceledException` within the legacy Event Hubs documentation where it is detailed as a service error related to disposed objects.  Because this occurs in our code only during cancellation, we're explicitly wrapping them to keep all cancellation scenarios unified to a `TaskCanceledException`